test/metabase/notification/send_test.clj

```diff
@@ -120,13 +124,218 @@
+                    retry/random-exponential-backoff-retry (constantly test-retry)]
+        (mt/with-temporary-setting-values [email-smtp-host "fake_smtp_host"
+                                           email-smtp-port 587]
+          (mt/reset-inbox!)
+          (#'notification.send/channel-send-retrying! 1 :notification/card {:channel_type :channel/email} fake-email-notification)
+          (is (= {:numberOfFailedCallsWithRetryAttempt 1}
+                 (get-positive-retry-metrics test-retry)))
+          (is (= 0 (count @mt/inbox)))))))
+  (testing "send email succeeds w/ retry"
+    (let [retry-config (assoc test-retry-configuration :max-attempts 2)
+          test-retry   (retry/random-exponential-backoff-retry "test-retry" retry-config)]
+      (with-redefs [email/send-email!                      (tu/works-after 1 mt/fake-inbox-email-fn)
+                    retry/random-exponential-backoff-retry (constantly test-retry)]
+        (mt/with-temporary-setting-values [email-smtp-host "fake_smtp_host"
+                                           email-smtp-port 587]
+          (mt/reset-inbox!)
+          (#'notification.send/channel-send-retrying! 1 :notification/card {:channel_type :channel/email} fake-email-notification)
+          (is (= {:numberOfSuccessfulCallsWithRetryAttempt 1}
+                 (get-positive-retry-metrics test-retry)))
+          (is (= 1 (count @mt/inbox))))))))
+
+(def ^:private fake-slack-notification
+  {:channel-id  "#test-channel"
+   :attachments [{:blocks [{:type "section", :text {:type "plain_text", :text ""}}]}]})
+
+(deftest slack-notification-retry-test
+  (notification.tu/with-send-notification-sync
+    (testing "post slack message succeeds w/o retry"
+      (let [test-retry (retry/random-exponential-backoff-retry "test-retry" test-retry-configuration)]
+        (with-redefs [retry/random-exponential-backoff-retry (constantly test-retry)
+                      slack/post-chat-message!               (constantly nil)]
+          (#'notification.send/channel-send-retrying! 1 :notification/card {:channel_type :channel/slack} fake-slack-notification)
+          (is (= {:numberOfSuccessfulCallsWithoutRetryAttempt 1}
+                 (get-positive-retry-metrics test-retry))))))
+    (testing "post slack message succeeds hiding token error"
+      (let [test-retry (retry/random-exponential-backoff-retry "test-retry" test-retry-configuration)]
+        (with-redefs [retry/random-exponential-backoff-retry (constantly test-retry)
+                      slack/post-chat-message!               (fn [& _]
+                                                               (throw (ex-info "Invalid token"
+                                                                               {:errors {:slack-token "Invalid token"}})))]
+          (#'notification.send/channel-send-retrying! 1 :notification/card {:channel_type :channel/slack} fake-slack-notification)
+          (is (= {:numberOfFailedCallsWithoutRetryAttempt 1}
```

Comment:
this was `:numberOfSuccessfulCallsWithoutRetryAttempt` before this PR, which is misleading.